File: binarySearchTree.ts
// import { inspect } from 'util'

class Node<T> {
  data: T
  left: Node<T>
  right: Node<T>

  constructor(data: T) {
    this.data = data
    this.left = null
    this.right = null
  }
}

/**
  Every BinarySearchTree have one root,
  from it the greater numbers will be on the right side and the smallers on the left.

  Exemplo:

                      right "12"
          right "9"
                      left "1"
  root "8"              
                      right "5"
          left "4" 
                      left "2"


  As seen above, even one element not being directly the root (begin),
  he can origin two more forks witch follow the same root's guide
 */
class BinarySearchTree<T = number> {
  root: Node<T>
  private lastRoot: Node<T>

  constructor() { }

  add(...data: Array<T>): void {
    let root = null
    if (this.lastRoot) {
      root = this.lastRoot
    }
    for (let index = 0; index < data.length; index++) {
      root = this.insert(root, data[index])
    }
    this.lastRoot = root
  }

  private insert(root: Node<T>, data: T) {
    if (!root) {
      this.root = new Node(data)
      return this.root
    }

    // smaller or equal -> left
    if (data <= root.data) {
      // if the left fork already exists, continue from it
      if (root.left) {
        return this.insert(root.left, data)
      } else {
        root.left = new Node(data)
      }
      // maior -> direita
    } else {
      // if the right fork alredy exits, continue from it
      if (root.right) {
        return this.insert(root.right, data)
      } else {
        root.right = new Node(data)
      }
    }

    return this.root
  }

  get getHeight() {
    return this.calculateHeight(this.root)
  }

  /**
    The height of a binarySearchTree is the distance between the root and its furthest leaf.
  */
  private calculateHeight(root: Node<T>) {
    let rightCount = 0
    let leftCount = 0

    if (!root) {
      return 0
    }

    if (root.left) {
      leftCount = this.calculateHeight(root.left) + 1
    }
    if (root.right) {
      rightCount = this.calculateHeight(root.right) + 1
    }

    return Math.max(rightCount, leftCount)
  }
}

/* Testing */
const tree = new BinarySearchTree()

let data = [3, 5, 2]
let data2 = [1, 4, 6, 7]

tree.add(...data)
tree.add(...data2)

// console.log(inspect(tree.root, { showHidden: false, depth: null }))
console.log(tree.getHeight) // -> 3

export { }

File: binarySearchTree.go
package main

import (
	"fmt"
)

func main() {
	bst := BinarySearchTree{}

	// bst.Insert(2)
	// bst.Insert(3)
	// bst.Insert(1)
	// bst.Insert(9)
	// bst.Insert(8)

	bst.Insert(30)
	bst.Insert(20)
	bst.Insert(25)
	bst.Insert(26)
	bst.Insert(21)

	// bst.Insert(30)
	bst.Delete(25)

	fmt.Println(bst.head.left.right.Value) // 26
	fmt.Println(bst.Search(26))            // true
	fmt.Println(bst.Search(25))            // false
}

type Node struct {
	Value int
	left  *Node
	right *Node
}

type BinarySearchTree struct {
	head *Node
}

func (b *BinarySearchTree) Delete(value int) {
	var previusNode *Node
	node := b.head
	for node != nil {
		if node.Value == value {
			if node.right != nil {
				l := node.right
				for l.left != nil {
					l = l.left
				}
				l.left = node.left
			}
			previusNode.right = node.right
			return
		}
		previusNode = node
		if value >= node.Value {
			node = node.right
		} else {
			node = node.left
		}
	}
}

func (b *BinarySearchTree) Search(value int) bool {
	node := b.head
	for node != nil {
		if node.Value == value {
			return true
		}
		if value >= node.Value {
			node = node.right
		} else {
			node = node.left
		}
	}
	return false
}

func (b *BinarySearchTree) Insert(value int) {
	if b.head == nil {
		b.head = &Node{
			Value: value,
		}
		return
	}

	node := b.head
	for node != nil {
		if value >= node.Value {
			// right
			if node.right != nil {
				node = node.right
			} else {
				node.right = &Node{
					Value: value,
				}
				break
			}
		} else {
			// left
			if node.left != nil {
				node = node.left
			} else {
				node.left = &Node{
					Value: value,
				}
				break
			}
		}
	}
}
